plugins/core-plugin/src/main/resources/Dockerfile-template-yaml

```diff
@@ -1,16 +1,67 @@
-FROM ${baseContainerImage}
+#===================================================================#
+# Create build environment from base Python template launcher image #
+#===================================================================#
+FROM ${basePythonContainerImage} as python-base
 
-# Copy template files to /template
+# Build args
 ARG WORKDIR=/template
-COPY main.py /template
-COPY requirements.txt /template
-COPY ${yamlTemplateName} /template/template.yaml
+ARG REQUIREMENTS_FILE=requirements.txt
+ARG BEAM_VERSION=${beamVersion}
+ARG BEAM_PACKAGE=apache-beam[dataframe,gcp,test,yaml]==$BEAM_VERSION
+
+# Copy template files to /template
+RUN mkdir -p $WORKDIR
+COPY main.py* requirements.txt* ${otherFiles} /template/
 WORKDIR $WORKDIR
 
-ENV FLEX_TEMPLATE_PYTHON_REQUIREMENTS_FILE=requirements.txt
+# Create requirements.txt file if not provided
+RUN if ! [ -f requirements.txt ] ; then echo "$BEAM_PACKAGE" > requirements.txt ; fi
+
+# Install dependencies to launch the pipeline and download to reduce startup time
+RUN python -m venv /venv \
+    && /venv/bin/pip install --no-cache-dir --upgrade pip setuptools \
+    && /venv/bin/pip install --no-cache-dir -U -r $REQUIREMENTS_FILE \
+    && /venv/bin/pip download --no-cache-dir --dest /tmp/dataflow-requirements-cache -r $REQUIREMENTS_FILE \
+    && rm -rf /usr/local/lib/python3.11/site-packages  \
+    && mv /venv/lib/python3.11/site-packages /usr/local/lib/python3.11/
+
+
+#============================================================#
+# Create Distroless xlang image compatible with YamlTemplate #
+#============================================================#
+FROM ${baseJavaContainerImage}
+
+# Build args
+ARG CHIPSET_ARCH=x86_64-linux-gnu
+ARG PYTHON_VERSION=3.11
+
+# Set python environment variables
 ENV FLEX_TEMPLATE_PYTHON_PY_FILE=main.py
 ENV PIP_NO_DEPS=True
 
+# Copy template, python wheels and python launcher script from python-base
+COPY --from=python-base /template /template
+COPY --from=python-base /tmp/dataflow-requirements-cache /tmp/dataflow-requirements-cache
+COPY --from=python-base /opt/google/dataflow/python_template_launcher /opt/google/dataflow/python_template_launcher
+
+# Copy python and installed packages from python-base
+COPY --from=python-base /usr/local/bin/python$PYTHON_VERSION /usr/local/bin/python
+COPY --from=python-base /usr/local/lib/python$PYTHON_VERSION /usr/local/lib/python$PYTHON_VERSION
+
+# Copy required shared libraries from python-base
```

Comment:
Where do these libraries come from?

Reply:
Generally, it feels like we're taking in base container images, but making a lot of assumptions about how they'll be configured. What constraints do we have on our base images here?

Reply:
They are being copied from Python base image which has a full install of python on it (with necessary licenses and libraries). I am copying the minimum shared c libraries that python uses to compile the graph since that is all that is done in the template launcher.

Ideally, there would be a staged distroless base image that has python, java and up-to-date libraries, but until that is possible, this provides significant speed up by copying just what is needed from both.

I don't know about "constraints" but the base image for python is only updated with the new version of beam, so nothing else should be changing other than the site-packages folder (which is not used by this anyway)

Reply:
I suppose the python base image could theoretically be kept static since it is just used as a build env, but using the parameter keeps it inline with the other templates that also use it (i.e. python templates and java templates with python UDF support), so that ti only needs to be kept updated in one place.

The java base image could also theoretically be kept static, but having a parameter for it allows a user to use a custom java image (albeit this use-case may not exist in the wild), and the same argument could be made for the python image too.

Reply:
I think we make too many assumptions about the structure of the base image to allow arbitrary images. For example `/tmp/dataflow-requirements-cache` is not guaranteed to exist (and won't exist) in most images

Reply:
We're also assuming that the java container doesn't have python already installed (may or may not be true). Generally, this dockerfile is fit pretty tightly to the images its building on, so I don't think this should be parameterized.

> so that ti only needs to be kept updated in one place.

We could still keep the version in a single place next to the images (we could even keep image names stripped of version and version as separate vars since I imagine generally people just want to update the version)

Reply:
The Java container does not have Python installed, it is Java distroless, so it doesn't have much more than a Java runtime and the template files/launcher.

`/tmp/dataflow-requirements-cache` is guaranteed to exist on python base image because that is where the python wheels are being downloaded to (also documented in https://cloud.google.com/dataflow/docs/guides/templates/configuring-flex-templates#supply-requirements-file)

I do get your point about it being tightly fitted, and while I think it would be nice that have a common definition so that they can stay in sync, I think there may be a few more iterations to this anyway, so I'll go ahead and hard code it for now.